### PR TITLE
[dag] message author and epoch checks on receiver side

### PR DESCRIPTION
**Pull request by ibalajiarun**
### Description

This PR adds the peer and author checks for the DAG message on the receiver side.

### CODE REVIEW

File: consensus/src/dag/dag_handler.rs [hunk 1]
```diff
@@ -47,15 +48,17 @@ impl NetworkHandler {
...
             }
         }
     }
 
     async fn process_rpc(&mut self, rpc_request: IncomingDAGRequest) -> anyhow::Result<()> {
-        let dag_message: DAGMessage = TConsensusMsg::from_network_message(rpc_request.req)?;
+        let dag_message: DAGMessage = rpc_request.req.try_into()?;
 
-        // TODO: verify epoch number and author
+        if dag_message.author()? != rpc_request.sender {
```
Comment (zekun000) [inline consensus/src/dag/dag_handler.rs:59]: not every message carries the author field, this basically bails all messages that don't have author field

**Review by zekun000 — CHANGES_REQUESTED** [review]

File: consensus/src/dag/dag_handler.rs (hunk 1 above)
Comment (ibalajiarun) [reply consensus/src/dag/dag_handler.rs:59]: true, here, we expect NodeMsg and CertifiedNodeMsg to have the author? any other message should still be an error?

Comment (zekun000) [reply consensus/src/dag/dag_handler.rs:59]: this is the main handler that every message will go through right? if we expect all rpc requests to have an author we should have more explicit error message like "unexpected rpc request"

Comment (ibalajiarun) [reply consensus/src/dag/dag_handler.rs]: Okay, I made the error explicit now.

File: consensus/src/dag/dag_handler.rs [hunk 2]
```diff
@@ -47,13 +48,20 @@ impl NetworkHandler {
...
     }
 
     async fn process_rpc(&mut self, rpc_request: IncomingDAGRequest) -> anyhow::Result<()> {
-        let dag_message: DAGMessage = TConsensusMsg::from_network_message(rpc_request.req)?;
+        let dag_message: DAGMessage = rpc_request.req.try_into()?;
+
+        let author = dag_message
+            .author()
+            .map_err(|_| anyhow::anyhow!("unexpected rpc message {:?}", dag_message))?;
+        if author != rpc_request.sender {
```
Comment (zekun000) [inline consensus/src/dag/dag_handler.rs:62]: we can still use ensure here

**Review by zekun000 — APPROVED** [review]

**Review by sasha8 — APPROVED** [review]

File: consensus/src/network.rs [hunk 3]
```diff
@@ -583,6 +585,18 @@ impl NetworkTask {
                             warn!(error = ?e, "aptos channel closed");
                         }
                     },
+                    ConsensusMsg::DAGMessage(request) => {
```
[the remark this answers is not in the document]
Comment (zekun000) [reply consensus/src/network.rs:588]: we need to wrap it with callbacks, but maybe we can just pass everything to epoch manager

[the remark this answers is not in the document]
Comment (sasha8) [reply consensus/src/network.rs:588]: Yes, this is what I was thinking.

File: consensus/src/dag/dag_handler.rs (hunk 2 above)
Comment (ibalajiarun) [reply consensus/src/dag/dag_handler.rs:62]: Yes, but ideally, we would log this here, so I kept this an if statement.

File: consensus/src/epoch_manager.rs [hunk 4]
```diff
@@ -1065,6 +1065,19 @@ impl EpochManager {
...
+                let dag_message = request.req;
+
+                if dag_message.epoch == self.epoch() {
+                    // TODO: send message to DAG handler
+                    Ok(())
+                } else {
+                    monitor!(
+                        "process_different_epoch_dag_rpc",
+                        self.process_different_epoch(dag_message.epoch, peer_id)
+                    )
```
[the remark this answers is not in the document]
Comment (ibalajiarun) [reply consensus/src/epoch_manager.rs:1078]: We dont treat QS message to decide epoch change. The block retrieval, I don't know.

**Pull request merged**